### ISSUE TITLE
Why a sigle method cannot be vesion neutral within a controller?

### ISSUE THREAD

**Issue by xperiandri** [Microsoft/aspnet-api-versioning#82]

**Comment by commonsensesoftware** [comment]
In a manner of speaking, they are.  At least with respect to these libraries, the thought process is `controller == service`; therefore, only the versioning of a service (aka API) can be influenced by routing.  At a method implementation level (aka action), you can bind or map an action for a particular version of the service. You can draw many parallels to how interface binding works in C#.  It's also important to remember that version-neutral means that any and all versions are accepted as opposed to not having a version.

When an action on a versioned controller does not have any attribution, then it maps (e.g. is implemented) by all API versions that the service defines. Once the `[MapToApiVersion]` attribute is applied to an action, it only matches the versions defined.  Under the hood, there is little difference between `[ApiVersion]` and `[MapToApiVersion]`. The reason for the difference is to make it very clear what the intention of the attribute decoration is.  The action level can map an API version, but it cannot declare a service-level API version.

Long story short, actions are intrinsically API version-neutral by default because they accept all API versions sent to the service, including if the service is API version-neutral itself.  An API-version neutral service will accept any specified version as well as no version. You can inspect the requested version API version via the `GetRequestedApiVersion()` extension method on the controller's request property in you want you own imperative logic within the controller.

For additional explanation, check out the [version interleaving](../wiki/Version-Interleaving) topic. I hope that helps clear things up.

**Comment by xperiandri** [comment]
Good point.

Then if I have a method for OpenId Connect User Info endpoint like this
``` C#
// GET: Account/UserInfo
/// <summary>
/// Returns additional information about the user.
/// </summary>
/// <returns>Additional information about the user.</returns>
/// <response code="200">JSON with user info returned.</response>
/// <response code="400">No passenger with specified id found.</response>
[Authorize(ActiveAuthenticationSchemes = OAuthValidationDefaults.AuthenticationScheme)]
[HttpGet("~/account/userinfo"), Produces("application/json")]
[ProducesResponseType(typeof(JObject), (int)HttpStatusCode.OK)]
[ProducesResponseType(typeof(OpenIdConnectResponse), (int)HttpStatusCode.BadRequest)]
public async Task<IActionResult> UserInfo()
```

I just need to change route to
``` C#
[HttpGet("/{version:apiVersion}/account/userinfo"), Produces("application/json")]
```

And if version 1 is the latest for now I have to specify it to OpenIddict as follows
``` C#
    .EnableUserinfoEndpoint("/1/account/userinfo")
```

Right?

**Comment by xperiandri** [comment]
It would be nice to align this with another endpoints which live in another controller that is marked as version neutral.
``` C#
    .EnableLogoutEndpoint("/connect/logout")
    .EnableTokenEndpoint("/connect/token")
```

**Comment by commonsensesoftware** [comment]
If you're just trying to have symmetry with the URLs, then you can totally use the route template you've defined and make the controller API version-neutral.  This will have the effect of requiring the client to specify an API version, but all versions will match your controller.

The API versioning extensions for OData do this very thing.  An extended, but API version-neutral MetadataController was added. The controller itself does not have any version-specific logic, but the requested API version does bind to the underlying Entity Data Model (EDM).  If I set up my routes the way you have, then I would end up with something like:

* ~/v1/$metadata
* ~/v2/$metadata
* ~/v3/$metadata
* ...

Currently, having a default API version for a URL that does not include the URL path segment is not supported. There is an interesting solution that someone in the community came up with that could enable that scenario.  I've taken it into consideration for a future release. For now, I think this approach gets you what you want.

**Comment by commonsensesoftware** [comment]
It would seem the current answer got you what you wanted or otherwise satisfied your needs. Feel free to reopen this issue or start a new thread as appropriate.
